### PR TITLE
[WIP][Experimet]  Fern SDK Experiment

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

An experiment in using the [fern SDK](https://github.com/fern-api/fern) to audo-generate a typescript & go client based on our OpenAPI spec.

Some hacks were added to make it work but the screenshots below show:
- All the functions are available via go-code
- This could potentially supersede #339 and involve less work in the future

Screenshot 2024-01-24 at 4 27 18 PM

Screenshot 2024-01-24 at 4 25 46 PM

## Issue

- #324 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by Olshansk** [comment]
Capturing a longer offline discussion here.

tl;dr Fern does not seem like the way to go.

---

After being asked by @ezeike what the purpose of fern is, I tried to summarize the motivation here:

```
In [1], we had to implement `GetAllApplications`. This involves things like:
- Writing & maintaining code
- Writing & maintaining docs
- Writing & maintaining tests
- Time spent on reviewing
- Needing to repeat this process for every additional function we need which is a proxy to anything in the OpenAPI spec
- Needing to repeat the process for every other language (e.g. typescript) we might need.

What I was thinking with [3] is that* we would:
- Get native language bindings for everything in [2] without writing code, docs or tests for all the functions
- Immediately have it be updated if any changes to the openapi spec happen
- Automatically get other languages it supports (e.g. typescript and python for free)

> For the SDK, move away from gRPC for GetApplications and use the Cosmos REST API for it instead?

This is where I may be missing something. The client generated provides a proto interface to req/res, so does it matter if we're using `gRPC` or a `REST API` with a golang native front?

[1] https://github.com/pokt-network/poktroll/pull/339
[2] https://raw.githubusercontent.com/pokt-network/poktroll/main/docs/static/openapi.yml
[3] https://github.com/pokt-network/poktroll/pull/344

 \* TBD if possible & where I need help and feedback on whether this experiment/idea is plausible
 ```
 
 Then @red-0ne found a few issues in the existing code:
 
Screenshot 2024-01-27 at 2 06 33 PM

I reached out to the team at fern on discord and was instructed to create two tickets for reference:

-  https://github.com/fern-api/fern/issues/2813#issuecomment-1912790978
-  https://github.com/fern-api/fern/issues/2814 

It turned out that the REST API exposed by the Cosmos SDK is not OpenAPI compliant. 

[In their documentation](https://docs.cosmos.network/main/learn/advanced/grpc_rest), they state:

>  gRPC-gateway REST Routes
> If, for various reasons, you cannot use gRPC (for example, you are building a web application, and browsers don't > support HTTP2 on which gRPC is built), then the Cosmos SDK offers REST routes via gRPC-gateway.

Since we are currently doing:
1. Use gRPC to create the OpenAPI spec (Cosmos default)
2. Use OpenAPI spec to generate golang SDK (using fern)
3. Hit issues because OpenAPI is non-compliant

The best path forward is to use gRPC and close out this experiment and go with the solution @ezeike put together here: https://github.com/pokt-network/poktroll/pull/339/
